abi.md

```diff
@@ -0,0 +1,147 @@
+
+###  Public Modules
+
+Any module considered a "public" module shall be implemented in Verilog in a
+consistent way.  Any public module shall exist as a Verilog module of the same
+name.
+
+Each public module with definitions (e.g. not external modules) shall be placed
+in a file with the same name.  No assumption is made of the filename of an
+implementation of an external module, it is the user's job to include files in
+such a way in their tools to resolve the name.
+
+### Port Lowering ABIv1
+
+Ports are generally lowered to netlist types, except where Verilog's type system
+prevents it.
+
+Ports of integer types shall be lowered to netlist ports (`wire`{.verilog}) as a
+packed vector of equivalent size.  For example, consider the following FIRRTL:
+
+```firrtl
+circuit Top :
+  module Top :
+    output out: UInt<16>
+    input b: UInt<32>
+```
+
+This is lowered to the following Verilog:
+
+```verilog
+module Top(
+    output wire [15:0] out,
+    input wire [31:0] in
+);
+```
+
+Ports of aggregate type shall be lowered according to the "Aggregate Type
+Lowering" description in the firrtl spec.
+
+Ports of enum type shall be lowered to ports of a type consistent with the rules
+for lowering firrtl enum types to Verilog.
+
+Ports of ref type shall be lowered to a Verilog macro of the form `` `define
+ref_<circuit name>_<module name>_<portname> <internal path from module>`` in a
+file with name `ref_<circuit name>_<module name>.sv`.
+
+### Port Lowering ABIv2
+
+Ports are lowered per the v1 ABI above, except for aggregate types.
+
+Vectors shall be lowered to Verilog packed vectors.
+
+Bundles shall be recursively split as per "Aggregate Type Lowering", except
+instead of recursively converting bundles to ground types, the recursion stops
+at passive types.
+
+Passive bundles shall be lowered to Verilog packed structs.
+
+Reference types in ports shall be logically split out from aggregates and named
+as though "Aggregate Type Lowering" was used.
+
+## On Types
+
+Types are only guaranteed to follow this lowering when the Verilog type is
+visible for some reason related to the visibility of the element.  These include
+use in ports and elements exported by reference through a public module.
+
+Ground types are lowered to the `logic`{.verilog} data type (SV.6.3.1), which is
+the 4-valued type.  It is important to distinguish the `logic`{.verilog} data
+type from using `logic`{.verilog} as a keyword to declare a variable instead of
+net-list object.
+
+Unsigned integers produce an unsigned packed bit vector.  Whether variable or
+netlist depends on the construct being used.
+
+Signed integers produce an unsigned packed bit vector.  Whether variable or
+netlist depends on the construct being used.
+
+Passive bundles, when lowered, are lowered to packed structs with their fields
+recursively following these lowering rules.
+
+Vectors, when lowered, are lowered to packed vectors with their element type
```

Comment:
It might be good hygiene to give explicit examples, in this section and earlier, of the FIRRTL types and the lowered SystemVerilog types that come from its lowering.

Reply:
Yes, but I think that can be a follow on PR.